File: Leetcode 2501-2600/2536/2536.py
# Leetcode 2536: Increment Submatrices by One
# https://leetcode.com/problems/increment-submatrices-by-one/
# Solved on 14th of November, 2025
class Solution:
    def rangeAddQueries(self, n: int, queries: list[list[int]]) -> list[list[int]]:
        """
        Applies a series of range increment queries to an n x n matrix.
        Each query specifies a submatrix to increment by one.

        Args:
            n (int): The dimension of the square matrix.
            queries (list[list[int]]): A list of queries, where each query is [r1, c1, r2, c2]
                                       representing the top-left (r1, c1) and bottom-right (r2, c2)
                                       coordinates of the submatrix to increment.

        Returns:
            list[list[int]]: The final n x n matrix after all queries have been applied.
        """
        diffMat = [[0] * (n + 1) for _ in range(n + 1)]

        for r1, c1, r2, c2 in queries:
            diffMat[r1][c1] += 1
            diffMat[r1][c2 + 1] -= 1
            diffMat[r2 + 1][c1] -= 1
            diffMat[r2 + 1][c2 + 1] += 1

        resultMat = [[0] * n for _ in range(n)]

        for r in range(n):
            rowSum = 0
            for c in range(n):
                rowSum += diffMat[r][c]
                if r > 0:
                    resultMat[r][c] = rowSum + resultMat[r - 1][c]
                else:
                    resultMat[r][c] = rowSum

        return resultMat

File: Leetcode 2501-2600/2536/2536.java
// Leetcode 2536: Increment Submatrices by One
// https://leetcode.com/problems/increment-submatrices-by-one/
// Solved on 14th of November, 2025
class Solution {
    /**
     * Applies a series of range addition queries to an n x n matrix.
     * @param n The size of the square matrix (n x n).
     * @param queries An array of queries, where each query is [r1, c1, r2, c2] representing a submatrix to increment by one.
     * @return The resulting matrix after applying all queries.
     */
    public int[][] rangeAddQueries(int n, int[][] queries) {
        int[][] diff = new int[n+1][n+1];
        for (int[] q : queries) {
            int r1 = q[0], c1 = q[1], r2 = q[2], c2 = q[3];
            diff[r1][c1] += 1;
            diff[r1][c2 + 1] -= 1;
            diff[r2 + 1][c1] -= 1;
            diff[r2 + 1][c2 + 1] += 1;
        }
        for (int i = 0; i <= n; i++) {
            for (int j = 0; j <= n; j++) {
                if (i > 0) {
                    diff[i][j] += diff[i - 1][j];
                }
                if (j > 0) {
                    diff[i][j] += diff[i][j - 1];
                }
                if (i > 0 && j > 0) {
                    diff[i][j] -= diff[i - 1][j - 1];
                }
            }
        }
        int[][] result = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                result[i][j] = diff[i][j];
            }
        }
        return result;
    }
}